File: SeventySix.Server/SeventySix.Application/Services/UserService.cs
// <copyright file="UserService.cs" company="SeventySix">
// Copyright (c) SeventySix. All rights reserved.
// </copyright>

using FluentValidation;
using SeventySix.Application.DTOs;
using SeventySix.Application.DTOs.Requests;
using SeventySix.Application.Extensions;
using SeventySix.Application.Interfaces;
using SeventySix.Domain.Interfaces;

namespace SeventySix.Application.Services;

/// <summary>
/// User service implementation.
/// Encapsulates business logic for user operations.
/// </summary>
/// <remarks>
/// This service implements the Service Layer pattern, providing a facade over
/// the repository layer while handling business logic and validation.
///
/// Design Principles Applied:
/// - Single Responsibility Principle (SRP): Focuses only on user business logic
/// - Dependency Inversion Principle (DIP): Depends on abstractions (IUserRepository, IValidator)
/// - Open/Closed Principle (OCP): Open for extension via dependency injection, closed for modification
///
/// Responsibilities:
/// - Validate requests using FluentValidation
/// - Coordinate between repository and mapping layers
/// - Enforce business rules
/// - Map between domain entities and DTOs
///
/// Transaction Management: Relies on repository layer for data persistence.
/// Validation: Uses FluentValidation for request validation before processing.
/// </remarks>
public class UserService : IUserService
{
	private readonly IUserRepository Repository;
	private readonly IValidator<CreateUserRequest> CreateValidator;

	/// <summary>
	/// Initializes a new instance of the <see cref="UserService"/> class.
	/// </summary>
	/// <param name="repository">The repository for data access operations.</param>
	/// <param name="createValidator">The validator for create requests.</param>
	/// <exception cref="ArgumentNullException">
	/// Thrown when repository or createValidator is null.
	/// </exception>
	/// <remarks>
	/// Dependencies are injected via constructor following Dependency Injection pattern.
	/// This enables loose coupling and facilitates unit testing with mocks.
	/// </remarks>
	public UserService(
		IUserRepository repository,
		IValidator<CreateUserRequest> createValidator)
	{
		Repository = repository ?? throw new ArgumentNullException(nameof(repository));
		CreateValidator = createValidator ?? throw new ArgumentNullException(nameof(createValidator));
	}

	/// <inheritdoc/>
	/// <remarks>
	/// Implementation retrieves all users from the repository and maps them to DTOs.
	/// Uses extension methods for clean entity-to-DTO transformation.
	/// </remarks>
	public async Task<IEnumerable<UserDto>> GetAllUsersAsync(CancellationToken cancellationToken = default)
	{
		var users = await Repository.GetAllAsync(cancellationToken);
		return users.ToDto();
	}

	/// <inheritdoc/>
	/// <remarks>
	/// Returns null if the user is not found, allowing the caller to determine
	/// the appropriate response (typically 404 Not Found).
	/// </remarks>
	public async Task<UserDto?> GetUserByIdAsync(int id, CancellationToken cancellationToken = default)
	{
		var user = await Repository.GetByIdAsync(id, cancellationToken);
		return user?.ToDto();
	}

	/// <inheritdoc/>
	/// <exception cref="ValidationException">
	/// Thrown when the request fails validation rules defined in CreateUserValidator.
	/// </exception>
	/// <remarks>
	/// Processing steps:
	/// 1. Validate request using FluentValidation (throws ValidationException if invalid)
	/// 2. Map request to domain entity using extension method
	/// 3. Persist entity via repository
	/// 4. Map created entity to DTO and return
	///
	/// The ValidateAndThrowAsync method ensures validation happens before any processing,
	/// following the fail-fast principle.
	/// </remarks>
	public async Task<UserDto> CreateUserAsync(
		CreateUserRequest request,
		CancellationToken cancellationToken = default)
	{
		// Validate request using FluentValidation
		await CreateValidator.ValidateAndThrowAsync(request, cancellationToken);

		// Map request to entity using extension method
		var user = request.ToEntity();

		// Save via repository
		var created = await Repository.CreateAsync(user, cancellationToken);

		// Map entity to DTO using extension method
		return created.ToDto();
	}
}

File: SeventySix.Server/SeventySix.Client/src/app/core/services/user.service.ts
/**
 * User service
 * Business logic layer for user operations
 */

import { inject, Injectable } from "@angular/core";
import { Observable } from "rxjs";
import { User, CreateUserRequest } from "@core/models/interfaces/user";
import { UserRepository } from "@core/repositories/user.repository";

/**
 * Service for user-related operations.
 * Implements business logic and delegates data access to repository.
 */
@Injectable({
	providedIn: "root"
})
export class UserService
{
	private readonly userRepository = inject(UserRepository);

	/**
	 * Get all users
	 */
	getAllUsers(): Observable<User[]>
	{
		return this.userRepository.getAll();
	}

	/**
	 * Get user by ID
	 * @param id User identifier
	 */
	getUserById(id: number): Observable<User>
	{
		return this.userRepository.getById(id);
	}

	/**
	 * Create new user
	 * @param request User creation request
	 */
	createUser(request: CreateUserRequest): Observable<User>
	{
		return this.userRepository.create(request);
	}

	/**
	 * Update existing user
	 * @param id User identifier
	 * @param user Partial user data to update
	 */
	updateUser(id: number, user: Partial<User>): Observable<User>
	{
		return this.userRepository.update(id, user);
	}

	/**
	 * Delete user
	 * @param id User identifier
	 */
	deleteUser(id: number): Observable<void>
	{
		return this.userRepository.delete(id);
	}
}
